### ISSUE TITLE
More docs and utilities on how to create custom resources

### ISSUE THREAD

**Issue by vimtor** [anomalyco/sst#5122]
The other day, I created a custom resource to deploy my online game to [Hathora](https://x.com/vimtor_/status/1860718951283499332). I got it working but needed to copy some utilities that are not exported from SST, such as converting `environment` and `link` into the `SST_RESOURCE_` environment variables.

As a newcomer to the Pulumi world, I was unsure whether to create an `sst.Resource` or a Pulumi component. I believe SST is incredibly powerful, and documenting this kind of thing would enable others to contribute their own providers and components.

**Comment by jayair** [comment]
Yup we want to make it easy to do this. Thanks for taking the time to dig into it.

**Comment by vimtor** [comment]
One thing that I found weird is that it is actually really hard to simply wrap an existing SDK. Pulumi has this thing called dynamic providers, but it gave me a lot of issues with SST. I ended up using the local command provider, but that isn't ideal. Having something like the command provider but that allowed you to pass a callback to run any code would be amazing.

**Comment by jl4nz** [comment]
I got stuck on this also, creating the resource and calling some aws sdk calls, I get:

```bash

|  Error
Error serializing '() => provider': index.js(50,43)

'() => provider': index.js(50,43): captured
  variable 'provider' which indirectly referenced
...
```

If the provider `async create()` function doesn't have any external function calls it works.

**Comment by vimtor** [comment]
> I got stuck on this also, creating the resource and calling some aws sdk calls, I get:
> 
> ```bash
> 
> |  Error
> Error serializing '() => provider': index.js(50,43)
> 
> '() => provider': index.js(50,43): captured
>   variable 'provider' which indirectly referenced
> ...
> ```
> 
> If the provider `async create()` function doesn't have any external function calls it works. 

I got the same issue when trying to create a Pulumi dynamic provider.

**Comment by ghandic** [comment]
Assuming y'all havent found a way around this?
